File: logs/lib/log_core.py
#!/usr/bin/env python3
"""
Core Logging Library for Python
Usage: from log_core import logger, log_event
"""

import json
import os
import re
from datetime import datetime, timezone
from pathlib import Path
from typing import Dict, Any, Optional

# Configure logs directory
LOGS_DIR = Path(os.environ.get('LOGS_DIR', Path(__file__).parent.parent / 'data' / 'logs'))
LOGS_DIR.mkdir(parents=True, exist_ok=True)

# Secret patterns to redact
SECRET_PATTERNS = [
    r'password',
    r'token',
    r'api_key',
    r'apikey',
    r'secret',
    r'credential',
    r'authorization',
    r'bearer',
]

def redact_secrets(data: str) -> str:
    """Redact sensitive information from log data."""
    for pattern in SECRET_PATTERNS:
        # JSON format: "key": "value"
        data = re.sub(
            rf'("{pattern}"[:\s]*")[^"]+',
            r'\1***REDACTED***',
            data,
            flags=re.IGNORECASE
        )
        # URL param format: key=value
        data = re.sub(
            rf'({pattern}[=\s]*)[^&\s,}}]+',
            r'\1***REDACTED***',
            data,
            flags=re.IGNORECASE
        )
    return data

def iso_timestamp() -> str:
    """Generate ISO 8601 timestamp."""
    return datetime.now(timezone.utc).strftime('%Y-%m-%dT%H:%M:%S.%f')[:-3] + 'Z'

def log_event(
    event_name: str,
    level: str = 'info',
    message: str = '',
    **extra_fields
) -> None:
    """
    Log a structured event to JSONL files.
    
    Args:
        event_name: Name of the event (e.g., 'user.login', 'api.call')
        level: Log level ('debug', 'info', 'warn', 'error', 'fatal')
        message: Human-readable message
        **extra_fields: Additional fields to include in the log entry
    """
    # Build event
    event = {
        'timestamp': iso_timestamp(),
        'event': event_name,
        'level': level,
        'message': message,
        **extra_fields
    }
    
    # Serialize to JSON
    event_json = json.dumps(event, separators=(',', ':'))
    
    # Redact secrets
    event_json = redact_secrets(event_json)
    
    # Write to per-event log
    event_log = LOGS_DIR / f'{event_name}.jsonl'
    with open(event_log, 'a') as f:
        f.write(event_json + '\n')
    
    # Mirror to unified stream
    all_log = LOGS_DIR / 'all.jsonl'
    with open(all_log, 'a') as f:
        f.write(event_json + '\n')

class Logger:
    """Structured logger with convenience methods."""
    
    def __init__(self, event_prefix: str = ''):
        self.event_prefix = event_prefix
    
    def _log(self, level: str, event_name: str, message: str, **kwargs):
        full_event = f'{self.event_prefix}.{event_name}' if self.event_prefix else event_name
        log_event(full_event, level, message, **kwargs)
    
    def debug(self, event_name: str, message: str, **kwargs):
        self._log('debug', event_name, message, **kwargs)
    
    def info(self, event_name: str, message: str, **kwargs):
        self._log('info', event_name, message, **kwargs)
    
    def warn(self, event_name: str, message: str, **kwargs):
        self._log('warn', event_name, message, **kwargs)
    
    def error(self, event_name: str, message: str, **kwargs):
        self._log('error', event_name, message, **kwargs)
    
    def fatal(self, event_name: str, message: str, **kwargs):
        self._log('fatal', event_name, message, **kwargs)

# Global logger instance
logger = Logger()

# Example usage
if __name__ == '__main__':
    logger.info('test', 'Test log entry', user_id=123, action='login')
    logger.error('test.failure', 'Test error', error_code=500, token='secret123')
    print(f'Logs written to {LOGS_DIR}')

File: logs/lib/log-core.js
/**
 * Core Logging Library for Node.js
 * Usage: const { logger, logEvent } = require('./log-core.js');
 */

const fs = require('fs');
const path = require('path');

// Configure logs directory
const LOGS_DIR = process.env.LOGS_DIR || path.join(__dirname, '..', 'data', 'logs');

// Ensure logs directory exists
if (!fs.existsSync(LOGS_DIR)) {
    fs.mkdirSync(LOGS_DIR, { recursive: true });
}

// Secret patterns to redact
const SECRET_PATTERNS = [
    /password/gi,
    /token/gi,
    /api_key/gi,
    /apikey/gi,
    /secret/gi,
    /credential/gi,
    /authorization/gi,
    /bearer/gi,
];

/**
 * Redact sensitive information from log data
 */
function redactSecrets(data) {
    let result = data;
    
    for (const pattern of SECRET_PATTERNS) {
        // JSON format: "key": "value"
        result = result.replace(
            new RegExp(`("${pattern.source}"[:\\s]*")[^"]+`, 'gi'),
            '$1***REDACTED***'
        );
        // URL param format: key=value
        result = result.replace(
            new RegExp(`(${pattern.source}[=\\s]*)[^&\\s,}]+`, 'gi'),
            '$1***REDACTED***'
        );
    }
    
    return result;
}

/**
 * Generate ISO 8601 timestamp
 */
function isoTimestamp() {
    return new Date().toISOString();
}

/**
 * Log a structured event to JSONL files
 */
function logEvent(eventName, level = 'info', message = '', extraFields = {}) {
    // Build event
    const event = {
        timestamp: isoTimestamp(),
        event: eventName,
        level: level,
        message: message,
        ...extraFields
    };
    
    // Serialize to JSON
    let eventJson = JSON.stringify(event);
    
    // Redact secrets
    eventJson = redactSecrets(eventJson);
    
    // Write to per-event log
    const eventLog = path.join(LOGS_DIR, `${eventName}.jsonl`);
    fs.appendFileSync(eventLog, eventJson + '\n');
    
    // Mirror to unified stream
    const allLog = path.join(LOGS_DIR, 'all.jsonl');
    fs.appendFileSync(allLog, eventJson + '\n');
}

/**
 * Structured logger with convenience methods
 */
class Logger {
    constructor(eventPrefix = '') {
        this.eventPrefix = eventPrefix;
    }
    
    _log(level, eventName, message, extraFields = {}) {
        const fullEvent = this.eventPrefix ? `${this.eventPrefix}.${eventName}` : eventName;
        logEvent(fullEvent, level, message, extraFields);
    }
    
    debug(eventName, message, extraFields = {}) {
        this._log('debug', eventName, message, extraFields);
    }
    
    info(eventName, message, extraFields = {}) {
        this._log('info', eventName, message, extraFields);
    }
    
    warn(eventName, message, extraFields = {}) {
        this._log('warn', eventName, message, extraFields);
    }
    
    error(eventName, message, extraFields = {}) {
        this._log('error', eventName, message, extraFields);
    }
    
    fatal(eventName, message, extraFields = {}) {
        this._log('fatal', eventName, message, extraFields);
    }
}

// Global logger instance
const logger = new Logger();

// Export
module.exports = {
    logEvent,
    logger,
    Logger,
    redactSecrets,
    isoTimestamp,
    LOGS_DIR
};

// Example usage
if (require.main === module) {
    logger.info('test', 'Test log entry', { userId: 123, action: 'login' });
    logger.error('test.failure', 'Test error', { errorCode: 500, token: 'secret123' });
    console.log(`Logs written to ${LOGS_DIR}`);
}
